narwhal/primary/src/primary.rs

```diff
@@ -852,33 +832,45 @@ impl PrimaryReceiverHandler {
             .map_err(DagError::StoreError)?;
 
         if let Some(vote_info) = result {
-            if header.epoch() < vote_info.epoch()
-                || (header.epoch() == vote_info.epoch() && header.round() < vote_info.round())
-            {
-                // Already voted on a newer Header for this publicKey.
-                return Err(DagError::TooOld(
-                    header.digest().into(),
+            ensure!(
+                header.epoch() == vote_info.epoch(),
+                DagError::InvalidEpoch {
+                    expected: header.epoch(),
+                    received: vote_info.epoch()
+                }
+            );
+            ensure!(
+                header.round() >= vote_info.round(),
+                DagError::AlreadyVotedNewerHeader(
```

Comment:
Let's also add this to the list of un-retriable errors https://github.com/MystenLabs/sui/blob/6a9abe9e38d9dc21156cc788860317170ff25838/narwhal/primary/src/primary.rs#L931